Graceful handling of unknown SAML id

When SAML sends a different ID for a given user and the one we have in dataverse, we would get the following error:

```
Caused by: java.lang.NullPointerException
        at edu.harvard.iq.dataverse.authorization.providers.saml.SamlConvertAccountPage.init(SamlConvertAccountPage.java:93)
```

We should handle this case gracefully and redirect the user the the failed login page. A log about the occurrence would also be helpful in the diagnosis of the issue.

--- a/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlAuthenticationServlet.java
+++ b/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlAuthenticationServlet.java
@@ -278,7 +278,7 @@ public class SamlAuthenticationServlet extends HttpServlet {
     private void redirectOnLoginIssue(HttpServletRequest request, HttpServletResponse response, SamlLoginIssue issue)
             throws IOException {
         request.getSession().setAttribute(SAML_LOGIN_ISSUE_SESSION_PARAM, issue);
-        response.sendRedirect("/failedLogin.xhtml");
+        response.sendRedirect(SamlFailedLoginPage.FAILED_LOGIN_XHTML);
     }
 
     private HttpServletRequest wrapHttpRequest(HttpServletRequest httpRequest) throws MalformedURLException {

--- a/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlConvertAccountPage.java
+++ b/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlConvertAccountPage.java
@@ -83,13 +83,15 @@ public class SamlConvertAccountPage implements Serializable {
         }
         userData = samlUser;
         if (!samlUser.isCompleteForLogin()) {
-            HttpSession session = JsfHelper.getCurrentSession();
-            session.setAttribute(SamlAuthenticationServlet.SAML_LOGIN_ISSUE_SESSION_PARAM,
-                    new SamlLoginIssue(SamlLoginIssue.Type.INCOMPLETE_DATA).addMessage(samlUser.printLoginData()));
-            return "/failedLogin.xhtml";
+            return failedLogin(SamlLoginIssue.Type.INCOMPLETE_DATA, samlUser);
         }
         AuthenticatedUser existingUser = authenticationService.getAuthenticatedUserByEmail(samlUser.getEmail());
         BuiltinUser existingBuiltinUser = builtinUserService.findByUserName(existingUser.getUserIdentifier());
+        if (existingBuiltinUser == null) {
+            logger.error("Authentication error. No builtin user found for {} with identifier {} (SAML id: {})",
+                    samlUser.getEmail(), existingUser.getUserIdentifier(), samlUser.getId());
+            return failedLogin(SamlLoginIssue.Type.AUTHENTICATION_ERROR, samlUser);
+        }
         builtinUsername = existingBuiltinUser.getUserName();
         consents = consentService.prepareConsentsForView(dataverseSession.getLocale());
         return StringUtils.EMPTY;
@@ -127,6 +129,13 @@ public class SamlConvertAccountPage implements Serializable {
         SamlUserData samlUser = (SamlUserData) session.getAttribute(SamlAuthenticationServlet.USER_TO_CONVERT_SESSION_PARAM);
         session.removeAttribute(SamlAuthenticationServlet.USER_TO_CONVERT_SESSION_PARAM);
         return samlUser;
+    }
+
+    private static String failedLogin(SamlLoginIssue.Type authenticationError, SamlUserData samlUser) {
+        HttpSession session = JsfHelper.getCurrentSession();
+        session.setAttribute(SamlAuthenticationServlet.SAML_LOGIN_ISSUE_SESSION_PARAM,
+                new SamlLoginIssue(authenticationError).addMessage(samlUser.printLoginData()));
+        return "/failedLogin.xhtml";
     }
 
     // -------------------- SETTERS --------------------

--- a/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlConvertAccountPage.java
+++ b/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlConvertAccountPage.java
@@ -135,7 +135,7 @@ public class SamlConvertAccountPage implements Serializable {
         HttpSession session = JsfHelper.getCurrentSession();
         session.setAttribute(SamlAuthenticationServlet.SAML_LOGIN_ISSUE_SESSION_PARAM,
                 new SamlLoginIssue(authenticationError).addMessage(samlUser.printLoginData()));
-        return "/failedLogin.xhtml";
+        return SamlFailedLoginPage.FAILED_LOGIN_XHTML;
     }
 
     // -------------------- SETTERS --------------------

--- a/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlFailedLoginPage.java
+++ b/dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/saml/SamlFailedLoginPage.java
@@ -16,6 +16,8 @@ import java.io.Serializable;
 @Named("FailedLogin")
 public class SamlFailedLoginPage implements Serializable {
 
+    public static String FAILED_LOGIN_XHTML = "/failedLogin.xhtml";
+
     private static final Logger logger = LoggerFactory.getLogger(SamlFailedLoginPage.class);
 
     private SamlLoginIssue samlLoginIssue;

